core/src/main/scala/shapeless/interface.scala

```diff
@@ -0,0 +1,424 @@
+  ): Tree = {
+
+    extractInterfaceInfo(weakTypeOf[T]) match {
+      case Left(NotTrait(nonTrait)) =>
+        if (c.compilerSettings contains "-Xlog-implicits") {
+          c.error(
+            nonTrait.typeSymbol.pos,
+            "Note: unsupported non-trait defined here"
+          )
+        } else ()
+        c.abort(
+          c.enclosingPosition,
+          "Unsupported Type: Interface-type generic represenation converters can only be generated for traits"
+        )
+      case Left(ByNameParameter(byNameParam)) =>
+        if (c.compilerSettings contains "-Xlog-implicits") {
+          c.error(
+            byNameParam.pos,
+            "Note: unsupported by-name parameter defined here"
+          )
+        } else ()
+        c.abort(
+          c.enclosingPosition,
+          "Unsupported Type: Interface-type generic represenation converters can only be generated for interfaces when none of their member methods have by-name parameters"
+        )
+      case Left(GenericMember(genericMember)) =>
+        if (c.compilerSettings contains "-Xlog-implicits") {
+          c.error(
+            genericMember.pos,
+            "Note: unsupported generic member defined here"
+          )
+        } else ()
+        c.abort(
+          c.enclosingPosition,
+          "Unsupported Type: Interface-type generic represenation converters can only be generated for traits with no generic members"
+        )
+      case Left(ConcreteMember(concreteMember)) =>
+        if (c.compilerSettings contains "-Xlog-implicits") {
+          c.error(
+            concreteMember.pos,
+            "Note: unsupported concrete member defined here"
+          )
+        } else ()
+        c.abort(
+          c.enclosingPosition,
+          "Unsupported Type: Interface-type generic represenation converters can only be generated for traits with no concrete members"
+        )
+      case Right(interfaceInfo) =>
+        val reprTypeTree = {
+          interfaceInfo.members.map {
+            case MemberInfo(name, paramLists, returnType) =>
+              val unifiedParamLists = {
+                paramLists
+                  .map(_.map(paramRepr).foldRight(tpeHNil)(tpeHCons))
+                  .foldRight(tpeHNil)(tpeHCons)
+              }
+              functionRepr(name)(tq"$unifiedParamLists => $returnType")
+          }.foldRight(tpeHNil)(tpeHCons)
+        }
+        val toGeneric = {
+          val target = TermName(c.freshName("target"))
+            val body = interfaceInfo.members.map {
+              case MemberInfo(name, paramLists, returnType) =>
+                val param = Ident(TermName(c.freshName("param")))
+                val arguments = {
+                  paramLists
+                    .zipWithIndex
+                    .map { case (paramList, listIndex) =>
+                      paramList.zipWithIndex.map { case (name, paramIndex) =>
+                        q"$param.apply($listIndex).apply($paramIndex)" }
```

Comment:
There is a hygiene problem hidden here. `param` is, as far as I understand, an `HList` of `HList`s. Is that correct? In that case, `apply` would require some implicits to resolve. Here's a safer version:

```diff
@@ -258,6 +241,10 @@
   def valueHCons(a: Tree, b: Tree): Tree = q"_root_.shapeless.::($a, $b)"
   val valueHNil: Tree = q"_root_.shapeless.HNil"
 
+  def selectByIndex(l: Tree, n: Int): Tree =
+    if (n == 0) q"$l.head"
+    else selectByIndex(q"$l.tail", n - 1)
+
   def materializeInterface[T: WeakTypeTag, R: WeakTypeTag](
     typeclassName: String,
     paramRepr: ParamInfo => Tree,
@@ -331,7 +318,8 @@
                     .zipWithIndex
                     .map { case (paramList, listIndex) =>
                       paramList.zipWithIndex.map { case (name, paramIndex) =>
-                        q"$param.apply($listIndex).apply($paramIndex)" }
+                        selectByIndex(selectByIndex(q"$param", listIndex), paramIndex)
+                      }
                     }
                 }
                 val paramType = {
@@ -361,7 +349,7 @@
                         q"${param.name}.asInstanceOf[$paramReprTpe]"
                       }.foldRight(valueHNil)(valueHCons)
                     }.foldRight(valueHNil)(valueHCons)
-                q"$target.apply($idx).apply($arg)"
+                q"${selectByIndex(q"$target", idx)}.apply($arg)"
               }
               val parameters = paramLists.map { paramList =>
                 paramList.map { param =>
```